haystack/nodes/other/join_docs.py

```diff
@@ -41,33 +48,38 @@ def __init__(
         self.top_k_join = top_k_join
 
     def run(self, inputs: List[dict], top_k_join: Optional[int] = None):  # type: ignore
-        if self.join_mode == "concatenate":
-            document_map = {}
-            for input_from_node in inputs:
-                for doc in input_from_node["documents"]:
-                    document_map[doc.id] = doc
-        elif self.join_mode == "merge":
-            document_map = {}
-            if self.weights:
-                weights = self.weights
-            else:
-                weights = [1 / len(inputs)] * len(inputs)
-            for input_from_node, weight in zip(inputs, weights):
-                for doc in input_from_node["documents"]:
-                    if document_map.get(doc.id):  # document already exists; update score
-                        document_map[doc.id].score += doc.score * weight
-                    else:  # add the document in map
-                        document_map[doc.id] = deepcopy(doc)
-                        document_map[doc.id].score *= weight
-        else:
-            raise Exception(f"Invalid join_mode: {self.join_mode}")
-
-        documents = sorted(document_map.values(), key=lambda d: d.score, reverse=True)
-
-        if top_k_join is None:
-            top_k_join = self.top_k_join
+        results = [inp["documents"] for inp in inputs]
+        weights = self.weights if self.weights else [1 / len(inputs)] * len(inputs)
+        document_map = {doc.id: [doc, 0] for doc in itertools.chain(*results)}
+
+        for result, weight in zip(results, weights):
+            for rank, doc in enumerate(result):
+                if self.join_mode == "concatenate":
+                    document_map[doc.id][1] = doc.score
+                elif self.join_mode == "merge":
+                    document_map[doc.id][1] += self._calculate_comb_sum(doc, weight)
+                elif self.join_mode == "reciprocal_rank_fusion":
+                    document_map[doc.id][1] += self._calculate_rrf(rank)
+                else:
+                    raise Exception(f"Invalid join_mode: {self.join_mode}")
 
+        sorted_docs = sorted(document_map.values(), key=lambda d: d[1], reverse=True)
+        docs = []
+        for document_score_pair in sorted_docs:
+            doc = document_score_pair[0]
+            doc.score = document_score_pair[1]
+            docs.append(doc)
+
+        top_k_join = top_k_join if top_k_join else self.top_k_join
         if top_k_join:
-            documents = documents[:top_k_join]
-        output = {"documents": documents, "labels": inputs[0].get("labels", None)}
+            docs = docs[:top_k_join]
+
+        output = {"documents": docs, "labels": inputs[0].get("labels", None)}
+
         return output, "output_1"
+
+    def _calculate_comb_sum(self, document, weight):
```

Comment:
You could put a minimal dosctring to remind us why it needs to be a separate function rather than an expression in the above method